server/src/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -1535,6 +1535,19 @@ private boolean upgradeRunningVirtualMachine(Long vmId, Long newServiceOfferingI
                     + ",memory=," + currentMemory + ")");
         }
 
+        _offeringDao.loadDetails(currentServiceOffering);
+        _offeringDao.loadDetails(newServiceOffering);
+
+        Map<String, String> currentDetails = currentServiceOffering.getDetails();
+        Map<String, String> newDetails = newServiceOffering.getDetails();
+        String currentVgpuType = currentDetails.get("vgpuType");
+        String newVgpuType = newDetails.get("vgpuType");
+        if(currentVgpuType != null) {
+            if(newVgpuType == null || !newVgpuType.equalsIgnoreCase(currentVgpuType)) {
+                throw new InvalidParameterValueException("Dynamic scaling of vGPU type is not supported. VM has vGPU Type: " + currentVgpuType);
+            }
+        }
+
         // Check resource limits
```

Comment:
Hi @anshul1886,
Could you extract the codes between lines 1541 - 1549 to a method with an auto described name, do some test cases to this new method and create a Javadoc explaining what the method do, what params it receive, the exceptions it throws and what can cause the exception?
Ty.